### PR TITLE
Implement XIP feature and enable ARC target support

### PR DESCRIPTION
**Pull request by wenyongh**

### CODE REVIEW

File: core/app-framework/app-native-shared/restful_utils.c [hunk 1]
```diff
@@ -100,41 +133,68 @@ request_t * unpack_request(char * packet, int size, request_t * request)
...
-char * pack_response(response_t *response, int * size)
+char *
+pack_response(response_t *response, int *size)
 {
     int len = RESPONSE_PACKET_FIX_PART_LEN + response->payload_len;
-    char * packet = (char*) WA_MALLOC(len);
-    if (packet == NULL)
+    uint16 u16;
+    uint32 u32;
+    char * packet;
```
Comment (no1wudi) [inline core/app-framework/app-native-shared/restful_utils.c:142]: char *packet

File: core/iwasm/compilation/aot_emit_aot_file.c [hunk 2]
```diff
@@ -2173,11 +2232,73 @@ aot_obj_data_create(AOTCompContext *comp_ctx)
...
+                LLVMDisposeMessage(err);
+                err = NULL;
+            }
+            aot_set_last_error("emit elf to object file failed.");
+            goto fail;
+        }
+
+        /* call arc gcc to compile assembly file to object file */
+        snprintf(buf, sizeof(buf), "%s%s%s%s%s%s",
+                 "/opt/zephyr-sdk/arc-zephyr-elf/bin/arc-zephyr-elf-gcc ",
```
Comment (xujuntwt95329) [inline core/iwasm/compilation/aot_emit_aot_file.c:2267]: Just a note: It should be better to read path from environment variable.

File: core/iwasm/compilation/aot_emit_aot_file.c [hunk 3]
```diff
@@ -2173,11 +2232,73 @@ aot_obj_data_create(AOTCompContext *comp_ctx)
...
+                 "-mcpu=arcem -o ", file_name, ".o -c ", file_name, ".s");
+        ret = system(buf);
+        /* remove temp assembly file */
+        snprintf(buf, sizeof(buf), "%s%s",file_name, ".s");
```
Comment (xujuntwt95329) [inline core/iwasm/compilation/aot_emit_aot_file.c:2271]:
```suggestion
        snprintf(buf, sizeof(buf), "%s%s", file_name, ".s");
```

File: core/app-mgr/app-manager/module_wasm_app.c [hunk 4]
```diff
@@ -653,7 +653,8 @@ wasm_app_module_install(request_t * msg)
                 AOT_SECTION_TYPE_FUNCTION,
                 AOT_SECTION_TYPE_EXPORT,
                 AOT_SECTION_TYPE_RELOCATION,
-                AOT_SECTION_TYPE_SIGANATURE
+                AOT_SECTION_TYPE_SIGANATURE,
```
Comment (xujuntwt95329) [inline core/app-mgr/app-manager/module_wasm_app.c:656]: I have a concern here:
If the signature section is not the last section, then is may be more difficult to implement the sign/verify logic

File: core/app-framework/app-native-shared/restful_utils.c [hunk 5]
```diff
@@ -63,35 +78,53 @@ char * pack_request(request_t *request, int * size)
...
         return NULL;
     }
     if (size < REQUEST_PACKET_FIX_PART_LEN) {
         return NULL;
     }
-    uint16 url_len = ntohs(*((uint16*) (packet + 12)));
-    uint32 payload_len = ntohl(*((uint32*) (packet + 14)));
 
-    if (size != ( REQUEST_PACKET_FIX_PART_LEN + url_len + payload_len)) {
+    memcpy(&u16, packet + 12, 2);
```
[the remark this answers is not in the document]
Comment (wenyongh) [reply core/app-framework/app-native-shared/restful_utils.c:100]: Yes, here is alignment issue found on ARC target.
Cannot use bh_memcpy_s as this file is shared for native and wasm app. When building wasm app, bh_memory_s isn't implemented.

File: core/app-mgr/app-manager/module_wasm_app.c (hunk 4 above)
Comment (wenyongh) [reply core/app-mgr/app-manager/module_wasm_app.c:656]: Here the array defines which sections to be destroyed after module is loaded to save the memory usage. As now signature section isn't generated by wamrc, in aot loader, we checks that the AOT file's last section must be relocation section (when custom section isn't generated) or custom section. When signature section is generated in the future, we can add more strict checks, e.g. the last section must be relocation section, custom section or signature section, and the order must be custom section < relocation section < signature section.

File: core/iwasm/compilation/aot_emit_aot_file.c (hunk 2 above)
Comment (wenyongh) [reply core/iwasm/compilation/aot_emit_aot_file.c:2267]: Yes, here had better get the path from environment variable firstly and check whether the path exists actually, and for following calling system() function, we had better use try..catch mechanism to handle the possible exceptions. I will add TODOs and let's fix them in the future.

File: product-mini/platforms/posix/main.c [hunk 6]
```diff
@@ -32,7 +32,9 @@ print_help()
     printf("  --stack-size=n         Set maximum stack size in bytes, default is 16 KB\n");
     printf("  --heap-size=n          Set maximum heap size in bytes, default is 16 KB\n");
     printf("  --repl                 Start a very simple REPL (read-eval-print-loop) mode\n"
-           "                         that runs commands in the form of `FUNC ARG...`\n");
+           "                         that runs commands in the form of \"FUNC ARG...\"\n");
+    printf("  --xip                  Enable XIP (Execution In Place) mode to run AOT file\n"
```
[the remark this answers is not in the document]
Comment (wenyongh) [reply product-mini/platforms/posix/main.c:36]: Do you mean passing the is_xip_mode flag to aot loader and aot loader exactly knows the mode? But it need to change the wasm_runtime_load() interface?

File: core/app-framework/app-native-shared/restful_utils.c [hunk 7]
```diff
@@ -32,29 +32,44 @@
...
+char *
+pack_request(request_t *request, int *size)
 {
     int url_len = strlen(request->url) + 1;
     int len = REQUEST_PACKET_FIX_PART_LEN + url_len + request->payload_len;
-    char * packet = (char*) WA_MALLOC(len);
-    if (packet == NULL)
+    uint16 u16;
+    uint32 u32;
+    char * packet;
```
[the remark this answers is not in the document]
Comment (wenyongh) [reply core/app-framework/app-native-shared/restful_utils.c]: Done

File: core/app-framework/app-native-shared/restful_utils.c (hunk 1 above)
Comment (wenyongh) [reply core/app-framework/app-native-shared/restful_utils.c]: Done

File: core/app-framework/app-native-shared/restful_utils.c [hunk 8]
```diff
@@ -32,29 +32,44 @@
 #define REQUES_PACKET_VER 1
 #define REQUEST_PACKET_FIX_PART_LEN 18
 #define REQUEST_PACKET_URL_OFFSET REQUEST_PACKET_FIX_PART_LEN
-#define REQUEST_PACKET_URL_LEN  *((uint16*)( (char*) buffer + 12)))   //!!! to ensure little endian
-#define REQUEST_PACKET_PAYLOAD_LEN *((uint32*)( (char*) buffer + 14)))   //!!! to ensure little endian
+#define REQUEST_PACKET_URL_LEN  *((uint16*)((char*) buffer + 12))   //!!! to ensure little endian
```
[the remark this answers is not in the document]
Comment (wenyongh) [reply core/app-framework/app-native-shared/restful_utils.c]: Done

File: core/app-framework/app-native-shared/restful_utils.c [hunk 9]
```diff
@@ -100,41 +133,68 @@ request_t * unpack_request(char * packet, int size, request_t * request)
...
     if (*packet != REQUES_PACKET_VER)
         return NULL;
+
     if (size < RESPONSE_PACKET_FIX_PART_LEN)
         return NULL;
-    uint32 payload_len = ntohl(*((uint32*) (packet + 12)));
+
+    memcpy(&u32, packet + 12, 4);
+    payload_len = ntohl(u32);
     if (size != ( RESPONSE_PACKET_FIX_PART_LEN + payload_len))
```
[the remark this answers is not in the document]
Comment (wenyongh) [reply core/app-framework/app-native-shared/restful_utils.c]: Done

File: core/iwasm/aot/aot_loader.c [hunk 10]
```diff
@@ -2208,21 +2312,69 @@ destroy_sections(AOTSection *section_list, bool destroy_aot_text)
...
+                    /* Read the count of native symbol */
+                    read_uint32(p, p_end, *p_count);
+                    return true;
+                }
+                p -= sizeof(uint32);
+            }
+        }
+        else if (section_type > AOT_SECTION_TYPE_SIGANATURE) {
+            set_error_buf(error_buf, error_buf_size,
+                          "resolve native symbol fail");
```
[the remark this answers is not in the document]
Comment (wenyongh) [reply core/iwasm/aot/aot_loader.c]: Done

File: core/iwasm/compilation/aot_emit_aot_file.c (hunk 3 above)
Comment (wenyongh) [reply core/iwasm/compilation/aot_emit_aot_file.c]: Done

File: product-mini/platforms/posix/main.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (no1wudi) [reply product-mini/platforms/posix/main.c:36]: I suspect that the aot loader can recognize the execution mode by number of native symbols.

[the remark this answers is not in the document]
Comment (wenyongh) [reply product-mini/platforms/posix/main.c:36]: Yes, as discussed previously, aot loader detects the execution mode by checking whether native symbol count is 0, if not, it doesn't mmap executable memory and apply relocation for text section, that's why here main.c must mmap executable memory for aot file to run it, if not, the memory is allocated by wasm_runtime_malloc and is not executable.

**Comment by xujuntwt95329** [comment]
LGTM

**Pull request merged**